RFC 0132 Monero Merge Mining (#2991)
tari-project/tari: RFC/src/RFC-0132_Merge_Mining_Monero.md

This satisfies **REQ 2**, proving that the Tari block was constructed before the parent chain's block. 

The block may now be mined. Once a solution is found that satisfies the Tari difficulty, the miner must include enough data to allow the Tari blockchain to assert **REQ 1** and **REQ 2**.
Concretely, A miner must include the following struct serialized using [bincode] in the `pow` field

```rust,ignore
pub struct MoneroPowData {
    /// Monero header fields
    pub header: MoneroBlockHeader,
    /// randomX vm key
    pub key: Vec<u8>,
    /// transaction count
    pub transaction_count: u16,
    /// transaction root
    pub transaction_root: [u8; 32],
    /// Coinbase merkle proof hashes
    pub coinbase_merkle_proof: Vec<[u8; 32]>,
    /// Coinbase tx from Monero
    pub coinbase_tx: MoneroTransaction,
}
```
_fig 2. Monero PoW data serialized in Tari blocks

A verifier may now check that the `coinbase_tx` contains the merge mining hash \\( h_m \\), and validate the `coinbase_merkle_proof` against the `transaction_root`.
The `coinbase_merkle_proof` contains the minimal proof required to construct the `transaction_root`.

> For example, a tree with a depth of 2 will have a proof size of 3 * 32 bytes.

Author: Need to update this to reflect the changes made in upcoming PR - 3 * 32 bytes + 2 u32s used for the depth and path_bitmap

```text
      [Root]*
      /   \
     h_c1  h2
    /    \
 [C]*     Tx1
 
 * Not included in proof
```

## Merge Mining Proxy 

The [Tari merge mining proxy] proxies the [Monero daemon RPC] interface. It behaves as a middleware that implements the merge 
mining protocol detailed above. This allows existing Monero miners to merge mine with Tari without having to make changes 
to mining software. 

The proxy must be configured to connect to a `monerod` instance, a Tari base node, and a Tari console wallet. Most requests

> are forwarded "as is" to `monerod`, however some are intercepted and augments before being returned to the miner.

Reviewer: ```suggestion
are forwarded "as is" to `monerod`, however some are intercepted and augmented before being returned to the miner.
```

shown here.

## Disclaimer

This document and its content are intended for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community of the
technological merits of the potential system outlined herein.

## Goals

This document describes the specific protocol Tari uses to merge mine with Monero.

## Related Requests for Comment

* [RFC-0131: Mining](RFC-0131_Mining.md)

## Introduction

Tari employs a hybrid mining strategy, accepting 2 mining algorithms whose difficulties are independent of each other as discussed in [RFC-0131_Mining.html](/RFC-0131_Mining.html).
This RFC details a protocol to enable Tari to accept Monero proof of work, enabling participating miners a chance to produce a valid block for either or both chain without additional mining effort.

The protocol must enable a Tari base node to make the following assertions:

**REQ 1.** The achieved mining difficulty exceeds target difficulty as dictated by Tari consensus, 

> **REQ 2.** The Monero block was constructed before the Tari block. This is to prevent a miner from submitting blocks from the parent chain that satisfy the auxiliary chain's difficulty without doing new work.

Reviewer 2: Perhaps also:
The Monero block was constructed after the previous Tari block

Author: Done

transactions included in the block. Additionally, a `blockhashing_blob` is a fixed size blob containing `serialized_monero_header`, `merkle_tree_root` and 
`txn_count` concatenated together. The `merkle_tree_root` is a _merkle root_ of the coinbase + the transaction hashes contained in the block.

```rust,ignore
pub struct Block {
    /// The block header
    pub header: BlockHeader,
    /// Coinbase transaction a.k.a miner transaction
    pub miner_tx: Transaction,
    /// References to the transactions included in this block
    pub tx_hashes: Vec<hash::Hash>,
}
```
_fig 1. The Monero block struct_

Next, modify the Monero block template by including the merge mining hash \\( h_m \\) in the extra fields of the coinbase transaction. Monero has a [merge mining subfield] 
to accommodate this data. Importantly, the extra field data part of the coinbase transaction hash and therefore the `merkle_tree_root`, the `blockhashing_blob` must be 
reconstructed. A rust port of [Monero's tree hash] algorithm is needed to achieve this. The coinbase hash MUST be the first element to be hashed when constructing the `merkle_tree_root`.

> This satisfies **REQ 2**, proving that the Tari block was constructed before the parent chain's block. 

Reviewer 2: Is the order correct here? Seems to contradict REQ 2

Author: Maybe this is better

```suggestion
This satisfies **REQ 2**, proving that the proof-of-work was performed for the Tari block.
```

Author: Done

The block may now be mined. Once a solution is found that satisfies the Tari difficulty, the miner must include enough data to allow the Tari blockchain to assert **REQ 1** and **REQ 2**.
Concretely, A miner must serialize `MoneroPowData` using Monero consensus encoding and add it to the `pow_data` field in the Tari header.

```rust,ignore
pub struct MoneroPowData {
    /// Monero header fields
    header: MoneroBlockHeader,
    /// randomX vm key
    randomx_key: FixedByteArray, // Fixed 64 bytes
    /// transaction count

>     transaction_count: u16,

Reviewer 2: Might be more compact to serialize this with VarInt's

Author: Will add some more details on the serialization and mention the VarInt encoding for all ints

Author: Added a Serialisation section

    /// transaction root
    transaction_root: Hash,
    /// Coinbase merkle proof hashes
    coinbase_merkle_proof: MerkleProof,
    /// Coinbase tx from Monero
    coinbase_tx: MoneroTransaction,
}
```
_fig 2. Monero PoW data serialized in Tari blocks

```rust,ignore
pub struct MerkleProof {
   branch: Vec<Hash>,

>    depth: u16,

Reviewer 2: These could also use VarInts

Author: Added a Serialisation section
